core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/http/timers/AsyncTimeoutTask.java

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright 2010-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.core.internal.http.timers;
+
+import java.util.concurrent.CompletableFuture;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.core.exception.SdkException;
+import software.amazon.awssdk.http.Abortable;
+import software.amazon.awssdk.utils.Logger;
+
+/**
+ * Asynchronous implementations of {@link TimeoutTask} to be scheduled to fail
+ * the {@link CompletableFuture} and abort the asynchronous requests.
+ */
+@SdkInternalApi
+public class AsyncTimeoutTask implements TimeoutTask {
+    private static final Logger log = Logger.loggerFor(AsyncTimeoutTask.class);
+    private final SdkException exception;
+    private volatile boolean hasExecuted;
+
+    private final CompletableFuture<?> completableFuture;
+    private Abortable abortable;
+
+    /**
+     * Constructs a new {@link AsyncTimeoutTask}.
+     *
+     * @param completableFuture the {@link CompletableFuture} to fail
+     * @param exception the exception to thrown
+     */
+    public AsyncTimeoutTask(CompletableFuture<?> completableFuture, SdkException exception) {
+        this.completableFuture = completableFuture;
+        this.exception = exception;
+    }
+
+    @Override
+    public void abortable(Abortable abortable) {
+        this.abortable = abortable;
+    }
+
+    @Override
+    public void run() {
+        log.trace(() -> "Times up. Aborting the request");
+        hasExecuted = true;
+        if (completableFuture != null && !completableFuture.isDone()) {
```

Comment:
When would we have a `null` future?

Reply:
yeah, it should not be null. I'll add the validation in the constructor.